packages/block-serialization-default-parser/src/index.js

```diff
@@ -0,0 +1,239 @@
+			}
+			return false;
+
+		case 'void-block':
+			// easy case is if we stumbled upon a void block
+			// in the top-level of the document
+			if ( 0 === stackDepth ) {
+				output.push( Block( blockName, attrs, [], '' ) );
+				offset = startOffset + tokenLength;
+				return true;
+			}
+
+			// otherwise we found an inner block
+			addInnerBlock(
+				Block( blockName, attrs, [], '' ),
+				startOffset,
+				tokenLength,
+			);
+			offset = startOffset + tokenLength;
+			return true;
+
+		case 'block-opener':
+			// we may have some HTML soup before the next block
+			const leadingHtmlStart = ( startOffset > offset ) ? offset : null;
+
+			// track all newly-opened blocks on the stack
+			stack.push(
+				Frame(
+					Block( blockName, attrs, [], '' ),
+					startOffset,
+					tokenLength,
+					startOffset + tokenLength,
+					leadingHtmlStart,
+				),
+			);
+			offset = startOffset + tokenLength;
+			return true;
+
+		case 'block-closer':
+			// if we're missing an opener we're in trouble
+			// This is an error
+			if ( 0 === stackDepth ) {
+				// we have options
+				//  - assume an implicit opener
+				//  - assume _this_ is the opener
+				//  - give up and close out the document
+				addFreeform();
+				return false;
+			}
+
+			// if we're not nesting then this is easy - close the block
+			if ( 1 === stackDepth ) {
+				addBlockFromStack( startOffset );
+				offset = startOffset + tokenLength;
+				return true;
+			}
+
+			// otherwise we're nested and we have to close out the current
+			// block and add it as a innerBlock to the parent
+			const stackTop = stack.pop();
+			stackTop.block.innerHTML += document.substr(
+				stackTop.prevOffset,
+				startOffset - stackTop.prevOffset,
+			);
+			stackTop.prevOffset = startOffset + tokenLength;
+
+			addInnerBlock(
+				stackTop.block,
+				stackTop.tokenStart,
+				stackTop.tokenLength,
+				startOffset + tokenLength,
+			);
+			offset = startOffset + tokenLength;
+			return true;
+
+		default:
+			// This is an error
+			addFreeform();
+			return false;
+	}
+}
+
+function nextToken() {
+	// aye the magic
+	// we're using a single RegExp to tokenize the block comment delimiters
+	// we're also using a trick here because the only difference between a
+	// block opener and a block closer is the leading `/` before `wp:` (and
+	// a closer has no attributes). we can trap them both and process the
+	// match back in Javascript to see which one it was.
+	const matches = tokenizer.exec( document );
+
+	// we have no more tokens
+	if ( null === matches ) {
+		return [ 'no-more-tokens' ];
+	}
+
+	const startedAt = matches.index;
+	const [ match, closerMatch, namespaceMatch, nameMatch, attrsMatch, voidMatch ] = matches;
+
+	const length = match.length;
+	const isCloser = !! closerMatch;
+	const isVoid = !! voidMatch;
+	const namespace = namespaceMatch || 'core/';
+	const name = namespace + nameMatch;
+	const hasAttrs = !! attrsMatch;
+	const attrs = hasAttrs ? JSON.parse( attrsMatch ) : null;
```

Comment:
I know there's a performance hit with `try`, but we should play it safe with `JSON.parse`, or generally speaking make sure we can inform the user of bad input and recover (e.g. isolate bad blocks) as best as possible. Thoughts, @dmsnell?

Reply:
There's no longer the famed V8 deoptimization with `try` / `catch`

https://github.com/petkaantonov/bluebird/wiki/Optimization-killers#2-unsupported-syntax
https://github.com/v8/v8/commit/9aac80f

Reply:
added the `try` in 9c85a60cf but left it out of the PHP since in PHP it already returns `null` on a failed parse